Adding ``qk_dag_topological_op_nodes`` to C API

alexanderivrii opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Addresses #15191.


### Details and comments

alexanderivrii committed:
adding C API methods qk_dag_num_instructions and qk_dag_topological_op_nodes

alexanderivrii committed:
removing unnecessary line

alexanderivrii committed:
Merge branch 'main' into c_dag_topo_op_nodes

alexanderivrii committed:
updating the PR; test; reno

eliarbel commented on crates/cext/src/dag.rs:
```diff
@@ -736,3 +736,62 @@ pub unsafe extern "C" fn qk_dag_free(dag: *mut DAGCircuit) {
+/// Return the operation nodes in the DAG listed in topological order.
+///
+/// @param dag A pointer to the DAG.
+/// @param order A pointer to an array of ``qk_dag_num_op_nodes(dag)`` elements
```
```suggestion
/// @param out_order A pointer to an array of ``qk_dag_num_op_nodes(dag)`` elements
```

alexanderivrii replied:
I have renamed `order` to `out_order`.

eliarbel commented on crates/cext/src/dag.rs:
```diff
@@ -736,3 +736,62 @@ pub unsafe extern "C" fn qk_dag_free(dag: *mut DAGCircuit) {
+/// qk_dag_topological_op_nodes(dag, order);
+///
+/// // do something with the ordered nodes
+/// for (uint32_t i=0; i<num_ops; i++) {
```
```suggestion
/// for (uint32_t i = 0; i < num_ops; i++) {
```
Just to have it comply with how `clang-format` would format it.

eliarbel commented on crates/cext/src/dag.rs:
```diff
@@ -736,3 +736,62 @@ pub unsafe extern "C" fn qk_dag_free(dag: *mut DAGCircuit) {
+///     printf("The gate at location %u is %u.\n", i, gate);
+/// }
+///
+/// // free the order array
```
```suggestion
/// // free the order array, register and dag pointers when done.
```

eliarbel commented on crates/cext/src/dag.rs:
```diff
@@ -736,3 +736,62 @@ pub unsafe extern "C" fn qk_dag_free(dag: *mut DAGCircuit) {
+/// consecutive elements of ``uint32_t``.
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_dag_topological_op_nodes(dag: *const DAGCircuit, order: *mut u32) {
```
```suggestion
pub unsafe extern "C" fn qk_dag_topological_op_nodes(dag: *const DAGCircuit, out_order: *mut u32) {
```
I think it's nicer to indicate by name that this is an output parameter.

jakelishman replied:
It's fairly normal in C APIs that pointers to `const` are implicitly "in" params and pointers to non-`const` are implicitly out. Either way, I don't think we've used an `out_` prefix in other cases we do this (e.g. the transpile result).

eliarbel replied:
The `const` vs non-`const` convention makes sense, but for the non-`const` cases we already have exceptions in the C API in the form of in-place modifies, like the standalone passes which change the input circuit, or the layout configuration setters, or the `*_free` functions which don't return a value at all. These cases are not the typical populate-user-buffers pattern. 
The `out_` convention is already used BTW in `qk_dag_op_node_gate_op`. So, whatever we decide to land on - we should make the API consistent in this regard. Let's decide on a convention so we can use it consistently.

jakelishman replied:
I didn't mention in-place, but that's still kind of part of the same conventions.  Really though, there's no one-size-fits-all, and in my experience, I've not seen C APIs that tried to do that.

I'm not really _against_ calling it `out_order`, I just think it's unnecessary and so wordier than it could be. I think there's somewhat a false split being made - as you mentioned with mutate-in-place parameters, "in" and "out" params aren't mutually exclusive. The `mut` alone makes it clear we'll be writing to it in some form, at which point we have to check the docs.

If you think it'd be much clearer as `out_order` that's ok by me.

alexanderivrii replied:
I have renamed `order` to `out_order` as per review comments (almost typed `out_of_order` instead).

eliarbel commented on crates/cext/src/dag.rs:
```diff
@@ -736,3 +736,62 @@ pub unsafe extern "C" fn qk_dag_free(dag: *mut DAGCircuit) {
+            .iter_mut()
+            .zip(out_topological_op_nodes)
+            .for_each(|(dest, src)| *dest = src.index() as u32);
+    };
```
```suggestion
    let out_slice = unsafe { std::slice::from_raw_parts_mut(out_order, dag.num_ops()) };
    out_slice
        .iter_mut()
        .zip(out_topological_op_nodes)
        .for_each(|(dest, src)| *dest = src.index() as u32);
```
It's more idiomatic to minimize the scope of `unsafe`.

eliarbel commented on releasenotes/notes/qk-dag-topological-op-nodes-0bf133675d81acbe.yaml:
```diff
@@ -0,0 +1,6 @@
+---
+features_c:
+  - |
+    Added ``qk_dag_topological_op_nodes`` to obtain the operation nodes in the dagcircuit
```
```suggestion
    Added :c:func:`qk_dag_topological_op_nodes` to obtain the operation nodes in the DAG circuit
```

eliarbel commented on test/c/test_dag.c:
```diff
@@ -312,6 +312,48 @@ static int test_dag_endpoint_node_value(void) {
+        printf("The number of op nodes %zu is not 0\n", num_ops);
+        qk_dag_free(dag);
+        qk_quantum_register_free(qr);
+        return EqualityError;
```
To better align with the memory deallocation convention in the C tests, you could replace this with something like: 
```c
    result = EqualityError;
    goto early_cleanup;
```
And modify the cleanup section below to: 
```c
cleanup:
    free(order);
early_cleanup:
    qk_dag_free(dag);
    qk_quantum_register_free(qr);
    return result;
```

eliarbel commented on test/c/test_dag.c:
```diff
@@ -312,6 +312,48 @@ static int test_dag_endpoint_node_value(void) {
+        if (order[1] != s_gate_idx) {
+            printf("Expected gate index %u but got %u\n", s_gate_idx, order[1]);
+            result = EqualityError;
+            goto cleanup;
```
```suggestion
```

eliarbel reviewed (commented):
Thanks Sasha for addressing this. Looks good. I've left some minor comments.

alexanderivrii committed:
Merge branch 'main' into c_dag_topo_op_nodes

alexanderivrii committed:
addressing review comments

alexanderivrii commented:
Thanks @eliarbel for the review. I have addressed your comments in [b171083](https://github.com/Qiskit/qiskit/pull/15297/commits/b1710836c2ddfbd334c0531a6289a67d9240bfcf), with the exception of renaming `order` to `out_order` (as per @jakelishman 's comment).

alexanderivrii committed:
fix

alexanderivrii committed:
Merge branch 'main' into c_dag_topo_op_nodes

alexanderivrii committed:
renaming order to out_order

jakelishman commented on crates/cext/src/dag.rs:
```diff
@@ -736,3 +736,60 @@ pub unsafe extern "C" fn qk_dag_free(dag: *mut DAGCircuit) {
+    out_slice
+        .iter_mut()
+        .zip(out_topological_op_nodes)
+        .for_each(|(dest, src)| *dest = src.index() as u32);
```
This technically isn't actually safe - Rust makes holding "safe" references to uninitialised data "undefined behaviour", and we expect the out pointer to point to uninitialised data in general.  We instead should probably do
```rust
for (i, node) in out_topological_op_nodes.enumerate() {
    // SAFETY: per documentation, `out_order` is aligned and points to a valid
    // aligned and maybe uninitialized block of memory valid for `num_op_nodes`
    // writes of `u32`s.
    unsafe { out_order.add(i).write(node.index() as u32) }
}
```
and never create `out_slice`.

alexanderivrii replied:
This is cool; and changed in [bcf4dee](https://github.com/Qiskit/qiskit/pull/15297/commits/bcf4deeba258c7db4cc0aa58a9c0200e5dbcb538). 

Note the previous way to do the copying was adapted (copied) from `qk_transpile_layout_initial_layout`, so we might want to change that code as well to avoid creating the `out_slice`.

eliarbel replied:
`qk_transpile_layout_output_permutation` and `qk_transpile_layout_final_layout` also use this pattern

eliarbel replied:
When I run the Rust layout tests with Miri, it doesn't complain about assigning values to the output `*mut u32` arrays using a slice. @jakelishman, mostly for education - any insight why?

jakelishman replied:
Ok, so I looked this up to make sure, and it's actually a slightly open question in Rust whether a `&[T]` is invalid if the values are uninitialised (see https://github.com/rust-lang/unsafe-code-guidelines/issues/412), which the current thinking being that it _won't_ be UB.  Consequently, Miri doesn't enforce "recursive validity" by default, but you can turn it on with `-Zmiri-recursive-validation`.

That said, imo we should still generally take care, and either do the direct pointer writing that I mentioned, or alternatively internally cast the pointer to `*mut MaybeUninit<T>` and then deal with `&[MaybeUninit<T>]` instead, to be 100% sure.  The examples we have here are fine, but if we ever get in the habit of using some `T: Drop` (e.g. taking an array of `CircuitData` or `DAGCircuit` or the like), then it _will_ be UB to assign to the slice directly, because Rust would attempt to execute the destructor on uninitialised memory.

---

Turning on the Miri flags, this is showing/explaining the error:

The assignment isn't the problem by itself, it's the creation of a `&[T]` where the `T` is uninitialized.  The Miri tests of `TranspileLayout` only pass initialised data in the `*mut result` pointers (since Rust makes it hard and unsafe to generate uninitialised data), so we wouldn't spot it even with `-Zmiri-recursive-validation` turned on.

The calling patterns from C of both that function (which should be changed, I agree) and this function are both
```c
uint32_t *order = malloc(sizeof(*order) * num_nodes);
qk_dag_topological_op_nodes(dag, order);
```
so `order` points to uninitialised data, and we shouldn't make slices over it.

You can get Miri to show the error with a script like this:
```rust
use std::{mem, slice};

unsafe fn write_8_slice(ptr: *mut u8) {
    let slice = slice::from_raw_parts_mut(ptr, 8);
    slice.fill(1);
}
unsafe fn write_8_direct(ptr: *mut u8) {
    (0..8).for_each(|i| unsafe { ptr.add(i).write(1) })
}

    
fn main() {
    let mut uninit = [mem::MaybeUninit::<u8>::uninit(); 8];
    // Fine.
    unsafe { write_8_direct(uninit.as_mut_slice().as_mut_ptr() as *mut u8)};

    let mut uninit = [mem::MaybeUninit::<u8>::uninit(); 8];
    // Miri complains.
    unsafe { write_8_slice(uninit.as_mut_slice().as_mut_ptr() as *mut u8)};
}
```
then run with
```bash
MIRIFLAGS='-Zmiri-recursive-validation' cargo +nightly miri run
```
```text
error: Undefined Behavior: constructing invalid value at .<deref>[0]: encountered uninitialized memory, but expected an integer
  --> src/main.rs:4:17
   |
 4 |     let slice = slice::from_raw_parts_mut(ptr, 8);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Undefined Behavior occurred here
   |
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
   = note: BACKTRACE:
   = note: inside `write_8_slice` at src/main.rs:4:17: 4:50
note: inside `main`
  --> src/main.rs:19:14
   |
19 |     unsafe { write_8_slice(uninit.as_mut_slice().as_mut_ptr() as *mut u8)};
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to 1 previous error
```

eliarbel replied:
Thanks Jake for the detailed analysis and explanation. I've opened an issue to track this in the other functions: https://github.com/Qiskit/qiskit/issues/15370

jakelishman commented on crates/cext/src/dag.rs:
```diff
@@ -736,3 +736,60 @@ pub unsafe extern "C" fn qk_dag_free(dag: *mut DAGCircuit) {
+/// qk_dag_add_quantum_register(dag, qr);
+///
+/// uint32_t qubit[1] = {0};
+/// qk_dag_apply_gate(dag, QkGate_H, qubit, NULL, false);
+/// qk_dag_apply_gate(dag, QkGate_S, qubit, NULL, false);
+///
+/// // get the number of operation nodes
+/// uint32_t num_ops = qk_dag_num_op_nodes(dag); // 2
+/// uint32_t *out_order = malloc(sizeof(uint32_t) * num_ops);
+///
+/// // get operation nodes listed in topological order
+/// qk_dag_topological_op_nodes(dag, out_order);
+///
```
Super minor style: you maybe want 3+ gates and to mix/match `front=True` and `front=False` so that the insertion order doesn't match the topological order.

alexanderivrii replied:
I definitely agree; and changed in [bcf4dee](https://github.com/Qiskit/qiskit/pull/15297/commits/bcf4deeba258c7db4cc0aa58a9c0200e5dbcb538). I think when I wrote the first draft of this PR the ability to add the nodes from the `front` was not yet available.

jakelishman commented on test/c/test_dag.c:
```diff
@@ -312,6 +312,46 @@ static int test_dag_endpoint_node_value(void) {
+
+    size_t num_ops = qk_dag_num_op_nodes(dag);
+    if (num_ops != 2) {
+        printf("The number of op nodes %zu is not 0\n", num_ops);
```
the number of op nodes _shouldn't_ be zero lol

alexanderivrii replied:
Fixed in [bcf4dee](https://github.com/Qiskit/qiskit/pull/15297/commits/bcf4deeba258c7db4cc0aa58a9c0200e5dbcb538).

jakelishman reviewed (commented):
tbh I preferred "order" over "out_order", but I don't feel hugely strongly.  It's not like it actually appears in user code anyway.

This looks about ready to go, thanks Sasha.

alexanderivrii committed:
merge with main branch

alexanderivrii committed:
review comments

alexanderivrii commented:
Regarding the never-ending ``order`` vs ``out_order`` kerfuffle, I am _almost_ tempted to commit the following code:

```rust    
pub unsafe extern "C" fn qk_dag_topological_op_nodes(dag: *const DAGCircuit, 
    #[cfg(beholder = "eli")] out_order: *mut u32,
    #[cfg(beholder = "jake")] order: *mut u32,
) {
...
}
```

eliarbel reviewed (approved):
Thanks Sasha for doing this and Jake for helping with the review.
